### PR TITLE
Use ansible to setup firewalld or iptables rules as appropriate

### PR DESCRIPTION
**Pull request by stbenjam**
I was running into an issue where firewalld was started and was
preempting the manual iptables configuration, with firewalld started I
wasn't able to reach the httpd server running on the provisioning host.
This ensures it's off and disabled.

We also weren't cleaning up the interfaces on RHEL, only CentOS.

### CODE REVIEW

**Comment by stbenjam** [comment]
Yea, you're right, I'll have to switch this to use firewalld for el8.  I think what caused my problems was I was doing metal3-dev-env, and switched to dev-scripts which enabled firewalld and broke metal3-dev-env.  I'll update this PR.

File: vm-setup/roles/firewall/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+# Use firewalld for RHEL 8
+- name: "Libvirt TCP ports"
+  firewalld:
+    zone: libvirt
+    port: "{{ item }}/tcp"
+    permanent: yes
+    state: enabled
+  loop: "{{ provisioning_tcp_ports + baremetal_tcp_ports }}"
+  become: true
+  when: rhel8
```
Comment (hardys) [inline vm-setup/roles/firewall/tasks/main.yml]: This aligns with what we've done historically, but I wonder if we could just use firewalld everywhere?  Or at least for RHEL and CentOS - it's running on my local CentOS 7.6 box

File: host_cleanup.sh [hunk 2]
```diff
@@ -43,7 +43,7 @@ ANSIBLE_FORCE_COLOR=true ansible-playbook \
     -b -vvv vm-setup/teardown-playbook.yml
 
 # There was a bug in this file, it may need to be recreated.
-if [[ $OS == "centos" ]]; then
+if [[ $OS == "centos" || $OS == "rhel" ]]; then
```
Comment (hardys) [inline host_cleanup.sh]: It might be worth breaking this into a separate PR - it's required for my WIP ipv6 work so we clean up the old ipv4 network-scripts

File: vm-setup/roles/firewall/defaults/main.yml [hunk 3]
```diff
@@ -0,0 +1,17 @@
+baremetal_tcp_ports: []
+baremetal_udp_ports:
+  - "6230"
+  - "6231"
+  - "6232"
+  - "6233"
+  - "6234"
+  - "6235"
```
Comment (hardys) [inline vm-setup/roles/firewall/defaults/main.yml]: We should probably set this as a fact generated based on num_masters and num_workers - see https://github.com/openshift-metal3/dev-scripts/pull/830

File: host_cleanup.sh (hunk 2 above)
Comment (stbenjam) [reply host_cleanup.sh]: https://github.com/metal3-io/metal3-dev-env/pull/95/files

File: vm-setup/roles/firewall/defaults/main.yml (hunk 3 above)
Comment (stbenjam) [reply vm-setup/roles/firewall/defaults/main.yml]: The range is configurable now, you'd be able to override this with `-e` in the ansible-playbook command 

```
VBMC_MAX_PORT=$((6230 + ${NUM_MASTERS} + ${NUM_WORKERS} - 1))
ansible-playbook [...]
   -e vmc_port_range="6230:$VBMC_MAX_PORT"
   [...]
```

File: vm-setup/roles/firewall/tasks/main.yml (hunk 1 above)
Comment (stbenjam) [reply vm-setup/roles/firewall/tasks/main.yml]: This all blows up on me, it seems like libvirt wants to use iptables, there's no libvirt zone on el7. Not sure why, needs more investigation I guess

**Comment by hardys** [comment]
centos CI failed with `FAIL - cluster-api-provider-baremetal-controller-manager statefulsets replicas correct expected 1, got null`

Not sure how we debug that without logs, will see if it reproduces locally

**Comment by stbenjam** [comment]
> centos CI failed with `FAIL - cluster-api-provider-baremetal-controller-manager statefulsets replicas correct expected 1, got null`
> 
> Not sure how we debug that without logs, will see if it reproduces locally

I had a good CentOS 7 test this morning, will kick off again, maybe a flake.

File: vm-setup/roles/firewall/defaults/main.yml [hunk 4]
```diff
@@ -0,0 +1,19 @@
+use_firewalld: true
+baremetal_interface: baremetal
+provisioning_interface: provisioning
+vbmc_port_range: "6230:6235"
```
Comment (hardys) [inline vm-setup/roles/firewall/defaults/main.yml]: I can fix it in a follow-up if you prefer, but as previously mentioned we should avoid hard-coding this, instead we can calculate the range based on the number of masters/workers, then generate a list of port numbers to iterate over, similar to https://github.com/metal3-io/metal3-dev-env/blob/master/vm-setup/roles/common/tasks/main.yml#L6

File: lib/common.sh [hunk 5]
```diff
@@ -108,6 +108,13 @@ if [[ ${os_version} -ne 7 ]] && [[ ${os_version} -ne 8 ]] && [[ ${os_version} -n
   exit 1
 fi
 
+# Use firewalld on CentOS/RHEL, iptables everywhere else
+export USE_FIREWALLD=False
+if [[ ($OS == "rhel" || $OS = "centos") && ${os_version} == 8 ]]
```
Comment (hardys) [inline lib/common.sh]: nit, we could do this logic in ansible since it's not used anywhere in the scripts now?

File: vm-setup/roles/firewall/defaults/main.yml (hunk 4 above)
Comment (stbenjam) [reply vm-setup/roles/firewall/defaults/main.yml]: I replied - https://github.com/metal3-io/metal3-dev-env/pull/91#discussion_r335995099

File: vm-setup/roles/firewall/tasks/main.yml (hunk 1 above)
Comment (hardys) [reply vm-setup/roles/firewall/tasks/main.yml]: Ah Ok, fair enough - I guess at some point we'll perhaps deprecate centos7 support and just standardize on CentOS/RHEL8

File: vm-setup/roles/firewall/defaults/main.yml (hunk 4 above)
Comment (hardys) [reply vm-setup/roles/firewall/defaults/main.yml]: Sorry I missed that, my point was that we're not overriding it with any calculated value, so for NUM_WORKERS of more than 3 the ports won't be opened.

I was thinking we could just pass the number of masters/workers in to ansible and calculate the port, but I could also do it in bash like I've proposed for dev-scripts.

Not a blocker, perhaps it's easier if we fix this as a follow-up since this looks ready to go otherwise and it's just maintaining the existing behaviour.

**Comment by hardys** [comment]
I tested this locally on RHEL8 and everything worked, including provisioning a node which was broken prior to this change.

The Centos CI is failing at `FAIL - centos-1 baremetal host reachable by ssh` but I can't see why - port 22 on my locally deployed node is definitely open

File: vm-setup/roles/firewall/defaults/main.yml (hunk 4 above)
Comment (stbenjam) [reply vm-setup/roles/firewall/defaults/main.yml]: Ah I see. I was hoping we could do it separately, just didn't want to change too much in one PR.

**Comment by stbenjam** [comment]
> I tested this locally on RHEL8 and everything worked, including provisioning a node which was broken prior to this change.
> 
> The Centos CI is failing at `FAIL - centos-1 baremetal host reachable by ssh` but I can't see why - port 22 on my locally deployed node is definitely open

I may have missed something in the iptables setup, CI runs on CentOS 7 I think.

**Comment by stbenjam** [comment]
/cc @maelk 

If you have a moment would you like sanity checking this as well?

File: vm-setup/roles/firewall/defaults/main.yml (hunk 4 above)
Comment (hardys) [reply vm-setup/roles/firewall/defaults/main.yml]: Sure that's fine, I can take care of it after this lands.

**Comment by maelk** [comment]
@stbenjam That looks like a flake in the CI to me. One of the machines get correctly provisioned but not the other. Considering the type of changes done, if there was an actual problem it would not work for either hosts. I'll give it another run

**Comment by hardys** [comment]
Hmm the centos job failed waiting on SSH again...

I tried pushing https://github.com/metal3-io/metal3-dev-env/pull/100 to validate the CI without any changes but it doesn't seem to be triggering a jenkins job - @maelk what are the triggers, do I need to remove the WIP label perhaps?

**Comment by maelk** [comment]
It looks like our centos test is broken today. We're looking into it. Meanwhile please discard the output of this job, @hardys the WIP label does not matter.

**Comment by hardys** [comment]
@maelk Ok thanks for confirming - @stbenjam do you want to go ahead and merge without CI or shall we wait?

We could also push a WIP pr to openshift-metal3/dev-scripts and prove the metal3 (CentOS based) CI there passes I guess?

**Comment by maelk** [comment]
It looks like Openstack fails to create the centos vms because of image corruption. we have a new image, trying again.

**Comment by hardys** [comment]
Ok centos CI passed - thanks @maelk :)

Are you happy to land this now - it lgtm
